### PR TITLE
Do not check XmlDocs unless required [WIP]

### PR DESCRIPTION
**Pull request by DedSec256**
Currently, XmlDoc check is performed even if the corresponding diagnostic (3390) is not enabled in the .fsproj.

This PR suggests adding an option to DiagnosticOptions to not call analysis if not required.

### CODE REVIEW

**Review by T-Gro — CHANGES_REQUESTED** [review]
Looks good in general, but it would be good to cover the functionality with tests covering the positive and negative cases.

**Comment by DedSec256** [comment]
@T-Gro, not sure if tests are needed here as this PR does not change the visible behavior in any way. The problem that this PR fixes is that checking XML docs is always done, even if its results are not required and therefore ignored. This PR allows analysis to be performed only when its results are required (i.e. `warnon:3390` option is enabled in .fsproj).

**Comment by T-Gro** [comment]
> @T-Gro, not sure if tests are needed here as this PR does not change the visible behavior in any way. The problem that this PR fixes is that checking XML docs is always done, even if its results are not required and therefore ignored. This PR allows analysis to be performed only when its results are required (i.e. `warnon:3390` option is enabled in .fsproj).

That is what I meant - if the warnon is enabled, a warning should still be produced.

**Comment by DedSec256** [comment]
There are a lot of tests for XmlDocs checking in
https://github.com/dotnet/fsharp/blob/main/tests/FSharp.Compiler.ComponentTests/Language/XmlComments.fs

with enabled 3390 diagnostic
https://github.com/dotnet/fsharp/blob/33f9cf87885fb1316023eca49a7bdc0c95e0812f/tests/FSharp.Test.Utilities/Compiler.fs#L433-L435

I will add one test that checks for the absence of warnings with disabled diagnostic

**Review by T-Gro — DISMISSED** [review]

File: src/Compiler/Facilities/DiagnosticOptions.fs [hunk 1]
```diff
@@ -31,3 +31,6 @@ type FSharpDiagnosticOptions =
             WarnAsError = []
             WarnAsWarn = []
         }
+
+    member x.CheckXmlDocs =
+        List.contains 3390 x.WarnOn && not (List.contains 3390 x.WarnOff)
```
Comment (0101) [inline src/Compiler/Facilities/DiagnosticOptions.fs:36]: Is this correct? [Docs for `warnon`](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/compiler-options) say "Enable specific warnings that might be off by default or **disabled by another command-line option**." So should we care about if it's in the `WarnOff` list?

Comment (DedSec256) [reply src/Compiler/Facilities/DiagnosticOptions.fs:36]: Currently, for warnings, it checks if they are not disabled even when they are enabled
https://github.com/dotnet/fsharp/blob/cecc728fb8aadc41647c24daed34a59186225acb/src/Compiler/Driver/CompilerDiagnostics.fs#L396-L400


However, if the warning is enabled and added to WarnAsError, then the rules are somewhat complicated, and they should also be taken into account, hm.
https://github.com/dotnet/fsharp/blob/cecc728fb8aadc41647c24daed34a59186225acb/src/Compiler/Driver/CompilerDiagnostics.fs#L419-L422

**Review by abonie — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Review by abonie — APPROVED** [review]
